講座一覧API改修 ※受講生側(DAIKI)

i-k-i-a-d opened the pull request:
やったこと
新たにAttendanceControllerを作成し講座一覧取得用のindexメソッドを作成し、CourseController内のindexメソッドを削除。
取得した講座情報を返すためのAttendanceIndexResourceクラスを作成し、CourseIndexResource.phpファイルを削除。
必要なくなったのでCourseIndexRequest.phpファイルを削除。
講座一覧取得用のルーティングを新たに作成し、元のルーティングを削除。
講座２のチャプター４，５にLessonを一つずつ加えて整合性を保つために関係する全てのSeederファイルを修正。

確認方法
Postmanで以下URLをSend。
http://localhost:8080/api/v1/attendance/index

すると以下戻り値が返ってくる。
{
    "data": [
        {
            "attendance_id": 1,
            "progress": 10,
            "course": {
                "course_id": 1,
                "title": "PHP入門講座",
                "image": "course/4459908b-3cdf-4521-94fa-c2a9746d92e1.png",
                "instructor": {
                    "instructor_id": 1,
                    "nick_name": "ニックネーム",
                    "last_name": "山田",
                    "first_name": "太郎",
                    "email": "test_instructor@example.com"
                }
            }
        },
        {
            "attendance_id": 2,
            "progress": 0,
            "course": {
                "course_id": 2,
                "title": "Laravel入門講座",
                "image": "course/dbe1f6ef-66b4-4ce0-bfef-7555b6213bd4.png",
                "instructor": {
                    "instructor_id": 2,
                    "nick_name": "ニックネーム2",
                    "last_name": "山田2",
                    "first_name": "太郎2",
                    "email": "test_instructor2@example.com"
                }
            }
        }
    ]
}

i-k-i-a-d committed:
新たにAttendanceControllerを作成し講座一覧取得用のindexメソッドを作成し、CourseController内のindexメソッドを削除。
取得した講座情報を返すためのAttendanceIndexResourceクラスを作成し、CourseIndexResource.phpファイルを削除。
必要なくなったのでCourseIndexRequest.phpファイルを削除。
講座一覧取得用のルーティングを新たに作成し、元のルーティングを削除。
講座２のチャプター４，５にLessonを一つずつ加えて整合性を保つために関係する全てのSeederファイルを修正。

kou-tech commented on database/seeds/LessonSeeder.php:
```diff
@@ -74,6 +74,26 @@ public function run()
                 'updated_at' => Carbon::now(),
                 'order'=> 1,
             ],
+            [
+                'chapter_id' => 4,
+                'url' => 'KH4MmQsCDux',
+                'title' => '概要',
+                'remarks' => "",
+                'status' => Lesson::STATUS_PUBLIC,
+                'created_at' => Carbon::now(),
+                'updated_at' => Carbon::now(),
+                'order'=> 1,
+            ],
+            [
+                'chapter_id' => 5,
+                'url' => 'KH4MmQsCDuy',
+                'title' => 'SVCモデルとは',
+                'remarks' => "",
+                'status' => Lesson::STATUS_PUBLIC,
+                'created_at' => Carbon::now(),
+                'updated_at' => Carbon::now(),
+                'order'=> 1,
+            ],
```
remarksはない場合、nullでいいと思います！
あとSVCではなく、MVCではないですかね...？

i-k-i-a-d replied:
完全に誤字ってました。これだと英語の第二文型ですね。
お恥ずかしい。

kou-tech commented on app/Http/Controllers/Api/Student/AttendanceController.php:
移植前の実装でリクエストのテキストで検索する機能も必要なのですが、その反映は以降のタスクに含める予定でしたでしょうか？

i-k-i-a-d replied:
そもそも、その機能が必要だと思ってませんでした。
実装済みです。`

kou-tech commented on app/Http/Controllers/Api/Student/AttendanceController.php:
```diff
@@ -0,0 +1,26 @@
+    public function index () {
+        $student = Auth::id();
+        $attendances = Attendance::with('course.instructor')->where('student_id', $student)->get();
+        return response()->json([
+            'data' => new AttendanceIndexResource($attendances)
+        ]);
```
振る舞いは変わらないのですが、次の記述で統一していましたので、修正をお願いします。

````php
return new AttendanceIndexResource($attendances);
````

i-k-i-a-d replied:
ResourceクラスにJSON形式に切り替える役割があったとは知りませんでした。

kou-tech commented on app/Http/Controllers/Api/Student/AttendanceController.php:
```diff
@@ -0,0 +1,26 @@
+     * @return AttendanceIndexResource
+     */
+    public function index () {
+        $student = Auth::id();
```
変数名は`$studentId`、あるいは`$authId`の方が良いと思います。

i-k-i-a-d replied:
確かに、生徒のidを取得するので、この方がわかりやすいですね。

kou-tech commented on app/Http/Resources/Student/AttendanceIndexResource.php:
移植元の実装は、コレクションメソッドの`mapメソッド`を使っていたと思います。
`foreach`は汎用的なループ構造のため、その過程でバグを踏みやすいです。
今回のようなデータの詰め替えのケースにおいては、`mapメソッド`を利用した実装に修正をお願いしたいです！

i-k-i-a-d replied:
一応、前任者のコードは確認していたのですが、mapメソッドの使い方がいまいちピンとこなかったので、自分でロジックをつくることにしたんですが、mapメソッドを使ってより限定的な記述方法を採用したほうがバグを防ぎやすいという考え方は理解しました。

kou-tech commented on app/Http/Controllers/Api/Student/AttendanceController.php:
```diff
@@ -0,0 +1,26 @@
+    /**
+     * 受講中講座一覧取得API
+     *
+     * @param void
```
ない場合はこの記述は削除して問題ないですが、本来は検索テキスト文字列がリクエストされます。

i-k-i-a-d replied:
検索バーから値を受取るという発想がなかったので削除していました。
今回、値を受取れるようにしたので作り直しました。

i-k-i-a-d committed:
LessonSeederのテストでーたのタイトルの一部を修正。
indexメソッドに検索バーから受け取った文字列を含むタイトルで講座情報を絞り込む機能を実装。
indexメソッドで講座ステータスがpublicとなっている講座情報だけに絞り込む機能を実装。
検索バーから受け取った値をバリデーションにかけるためにリクエストクラスを実装。
Resourceクラスで新たにコレクションを組み直すロジックをmapメソッドで書き替え。
講座ステータスがpublicとなるデータの抽出はindexメソッド内で完結するのでextractPublicCourseメソッドは削除。

kou-tech commented on app/Http/Controllers/Api/Student/AttendanceController.php:
```diff
@@ -0,0 +1,38 @@
+     */
+    public function index (AttendanceIndexRequest $request) {
+        $studentId = Auth::id();
+        $attendances = Attendance::with('course.instructor')->where('student_id', $studentId)->whereHas('course', function (Builder $query) {
+            $query->where('status', Course::STATUS_PUBLIC);
+        })->get();
+
+        if (!$request->search_word) {
+            return new AttendanceIndexResource($attendances);
+        }   
+
+        $attendances = Attendance::with('course.instructor')->where('student_id', $studentId)->whereHas('course', function (Builder $query) use($request) {
+            $query->where('title', 'like', "%{$request->search_word}%");
+            $query->where('status', Course::STATUS_PUBLIC);
+        })->get();
+
+        return new AttendanceIndexResource($attendances);
```
これだと検索ワードがある場合に、2回DBにアクセスしてしまうのでif内に閉じ込めた方がよいです！

````php
        if (!$request->text) {
            $attendances = Attendance::with('course.instructor')
            ->where('student_id', $studentId)
            ->whereHas('course', function (Builder $query) {
                $query->where('status', Course::STATUS_PUBLIC);
            })
            ->get();
            return new AttendanceIndexResource($attendances);
        }
````

kou-tech replied:
クエリ部分がメソッドチェーンが多くなり、横に広すぎて可読性が落ちているので、上のコメントのように改行をお願いしたいです！

i-k-i-a-d replied:
確かに、言われてみれば、検索ワードがあるときも検索ワードがないときの処理をしてしまってました。反省します。

改行を入れてもいいのは知りませんでした。
でも、改行を入れていい場所と悪い場所がありそうですが、よかったら教えていただけれるとありがたいです。

kou-tech replied:
@i-k-i-a-d 
> 改行を入れていい場所と悪い場所がありそう

構文解析できないレベルはエラーとなるので気付けるかと思います。
次のような場合です。

````php
Attendance:
:with('course.instructor')
````

ただそれ以外で言うと、現場のコードルールや好み、Laravelの標準ルールがあったりするので、明確な答えはありませんでした...。私の場合はコード整形ツールを現場で導入しているので、そちらに従っています。

i-k-i-a-d committed:
検索ワードがあるときとないときで、1回ずつDBへアクセスするように条件分岐のロジックを修正。
可読性を高めるために適所に改行を入れた。
